_episodes/01-why-bother-with-performance.md

```diff
@@ -0,0 +1,235 @@
+teaching: 10
+exercises: 5
+questions:
+- "What is software performance?"
+- "Why is software performance important?"
+- "How can performance be measured?"
+- "What is meant by flops, walltime and cpuh?"
+- "What affects performance?"
+objectives:
+- "Understand the necessity of code optimisation"
+- "Identify the different methods of enhancing performance"
+- "Calculate walltime, cpuh"
+keypoints:
+- "Understanding performance is the best way of utilising your HPC resources efficiently"
+- "There are many ways of enhancing performance"
+- "Just because one way is correct doesn't always mean it is the best option"
+---
+
+## What is software performance?
+
+Before getting into the software side of things, lets take a few steps back. The concept
+of performance is generic and can apply to many factors in our lives, such as our own
+physical and mental performance. With *software performance* the emphasis is not on the
+using the most powerful machine, but on how best to utilise the power that you have.
+
+Say you are the chef in a restaurant and every dish that you do is perfect. You would be
+able to produce a set 7 course meal for a table of 3-6 without too much difficulty. If
+you catering a conference dinner though, it becomes more difficult, as people would be
+waiting many hours for their food. However, if you could delegate tasks to a team of 6
+additional chefs who could assist you (while also communicating with each other), you
+have a much higher chance of getting the food out on time and coping with a large
+workload. With 7 courses and a total of 7 chefs, it's most likely that each chef will
+spend most of their time concentrating on doing one course.
+
+When we dramatically increase the workload we need to distribute it *efficiently* over
+the resources we have available so that we can get the desired result in a reasonable
+amount of time. That is the essence of software performance, using the resources you
+have to the best of their capabilities.
+
+## Why is software performance important?
+
+This is potentially a complex question but has a pretty simple answer when we restrict
+ourselves to the context of this lesson. Since the focus here is our *usage* of
+software that is developed by someone else, we can take a self-centred and simplistic
+approach: all we care about is **reducing the time to solution** to an acceptable level
+while **minimising our resource usage**.
+
+This lesson is about taking a well-informed, systematic approach on how to do this.
+
+> ## Enchancing performance: rationale
+>
+> Imagine you had a `10x10x10` box like the one below, divided up into smaller boxes,
+> each measuring `1x1x1`. In one hour, one CPU core can simulate one hour of activity
+> inside the smaller box. If you wanted to simulate what was happening inside the large
+> box for 8 hours, how long will the example below take to run if we only use one CPU
+> core?
+>
+> ![10x10x10_cube](../fig/01/10x10x10_cube.png)
```

Comment:
@cilwerner A 40^3 box is going to be impossible to see on a screen, let's make a smaller one that should be clearer